### PR TITLE
Better Life Engine with Shocking Relative/Stranger Mechanics and Variety

### PR DESCRIPTION
**Pull request by M4YX0R**
Variety Ratio (%) = Species/Population*100

It provides a better depth of information about variety than the species data.

Variety Ratio 100%: All organisms are different.
Variety Ratio 50%: Each one of the 2 organisms is similar to the other. Each species has an average of 2 individuals.
Variety Ratio 25%: Each one of the 4 organisms is similar to the other 3. Each species has an average of 4 individuals.
Variety Ratio 10%: Each one of the 10 organisms is similar to the other 9. Each species has an average of 10 individuals.
Variety Ratio 0%: All organisms are the same.

AND THE GREAT SHOCK!

Organisms now carry identities that express kinship ties by increasing the id of their parents by one. Just as this.id+1 represents the child's id and this.id-1 represents the parent's id, the one with the id value of this.id-2 is the grandfather and the one with this.id+2 is the grandchild. Id values ​​are not unique and are only set to express kinship ties. Thus, if this.id is equal to the Id of any other organism seen, we have come across a cousin or sibling. In this way, siblings and cousins ​​are also considered in kinship.

Establishing kinship ties via Id and only comparing numbers for kinship ties ensures that there is no disadvantage in terms of performance.

It is classified according to whether the visible organism is related or foreign, with a generation tolerance that can be determined by the user. In the brain, the catch-or-flight response to kinship is evolving. This leads to the miraculous evolutions I stumbled upon during testing below!

- As expected, in an environment where kill cells have evolved, hunters have developed the behavior of "get away from the relative" and "catch the stranger" in order not to harm their relatives.

- In order to reach the maximum number of food in a crowded environment where kill cells have not yet evolved, organisms have developed an "escape from kin" behavior, thus gaining a huge evolutionary advantage by eating the food of strangers and leaving that of relatives!

- In order to increase diversity and prevent extinction, an attempt was made to evolve a plant-animal mutual relationship in an environment with abundant walls. This balance, which was previously only achieved by armor cells, has now resulted in farmer organisms harvesting only the food seen with the "escape from the stranger" behavior, despite the kill cells!

THIS IS SHOCKING!
- In an environment where food is high, I suddenly reduced the amount of food and quickly the organisms began to starve. Meanwhile, organisms evolved something miraculous! Organisms that evolved the "catch kin" behavior altruistically went to their closest relatives and died alongside their relatives before starving to death. Thus, his relatives survived by feeding on his corpse and passed on at least some of his genetics to future generations!

I should also mention that I am trying to develop a "selective kill" for numerous other possibilities, such as ganging up with relatives. I will share it soon.

### CODE REVIEW

File: src/Organism/Perception/Brain.js [hunk 1]
```diff
@@ -30,24 +30,31 @@ class Brain {
...
     }
-
+    
     observe(observation) {
         this.observations.push(observation);
     }
-
+    
     decide() {
+        //Generation difference count: Eg(3): -3(father of grandfather) to +3(son of grandson)
```
Comment (M4YX0R) [inline src/Organism/Perception/Brain.js:53]: this.id-3 = id for the father of the grandfather
this.id-2 = id for the grandfather
this.id-1 = id for the father
this.id-0 = id for cousins and siblings
this.id+1 = id for the son
this.id+2 = id for the grandson
this.id+3 = id for the son of the grandson

**Comment by doo158** [comment]
i am assuming this is something you have actually made. i'm not familiar with github, is there a way for me to play this myself?

**Comment by FireAndIceFrog** [comment]
> 

Just checkout the branch and you should be able to play with it :)

File: src/Organism/Perception/Brain.js (hunk 1 above)
Comment (xkortex) [reply src/Organism/Perception/Brain.js:53]: I don't think IDs are guaranteed to be monotonic, nor should they be. It's not clear how the ID maps to kinship relations.

**Comment by xkortex** [comment]
Note: I'm not the lead dev. 

I like the overall idea. I'm not sold on the ID mechanism with linear relationships. Family is quite famously a tree structure. I think you'd be better off with some sort of numerical structure such as a bitmask where you randomly edit bits during mutation, and the kinship is measured by the edit distance. Not unlike how real genetic similarity works. 

Regardless of the mechanism, it should probably be operating in the Species data structure level, for performance reasons.

**Comment by MaxRobinsonTheGreat** [comment]
I am closing PRs I consider mods and moving them to the discord server

**Pull request closed without merging**